lib/ferrum_pdf.rb

```diff
@@ -13,23 +13,39 @@ module FerrumPdf
   autoload :Controller, "ferrum_pdf/controller"
   autoload :HTMLPreprocessor, "ferrum_pdf/html_preprocessor"
 
+  mattr_accessor :browser_mutex, default: Mutex.new
   mattr_accessor :include_assets_helper_module, default: true
   mattr_accessor :include_controller_module, default: true
-  mattr_accessor :browsers, default: {}
   mattr_accessor :config, default: ActiveSupport::OrderedOptions.new.merge(
     window_size: [ 1920, 1080 ]
   )
 
+  # This doesn't use mattr_accessor because having a `.browser` getter and also
+  # having local variables named `browser` would be confusing. For simplicity,
+  # this is explicitly accessed.
+  @@browser = nil
+
   class << self
     def configure
       yield config
     end
 
-    # Creates and registers a new browser for exports
-    # If a browser with the same name already exists, it will be shut down before instantiating the new one
-    def add_browser(name, **options)
-      @@browsers[name].quit if @@browsers[name]
-      @@browsers[name] = Ferrum::Browser.new(@@config.merge(options))
+    # Sets the browser instance to use for all operations
+    # If a browser is already set, it will be shut down before setting the new one
+    def browser=(browser_instance)
+      browser_mutex.synchronize do
+        @@browser&.quit
+        @@browser = browser_instance
+      end
+    end
+
+    # Provides thread-safe access to the browser instance
+    def with_browser
+      browser_mutex.synchronize do
+        @@browser ||= Ferrum::Browser.new(config)
+        @@browser.restart unless @@browser.client.present?
+        yield @@browser
+      end
```

Comment:
```suggestion
    def with_browser(browser=nil)
      if browser.present?
        yield browser
      else
        browser_mutex.synchronize do
          @@browser ||= Ferrum::Browser.new(config)
          @@browser.restart unless @@browser.client.present?
          yield @@browser
        end
      end
```

If we want to support the `browser: nil` option for `render_pdf` to let users pass in their own separate browser, we could do something like this.